roachtest: out-of-bounds error in rangeStatsOperator

roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6315043?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6315043?buildTab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [b316a5ed5fe7253d113174d9d95ddebf1143b4e4](https://github.com/cockroachdb/cockroach/commits/b316a5ed5fe7253d113174d9d95ddebf1143b4e4):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:264,sqlsmith.go:324,test_runner.go:897: error: pq: internal error: unexpected error from the vectorized engine: runtime error: index out of range [5] with length 5
		stmt:
		SELECT
			'2018-12-27 04:49:18.000437+00:00':::TIMESTAMPTZ AS col_283066,
			tab_115474.crdb_internal_mvcc_timestamp AS col_283067,
			crdb_internal.range_stats(tab_115474._bytes::BYTES)::JSONB AS col_283068,
			true AS col_283069,
			tab_115474.crdb_internal_mvcc_timestamp AS col_283070
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_115474
		WHERE
			tab_115474._bool
		LIMIT
			5:::INT8;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-19234

--- a/pkg/sql/colexec/range_stats.go
+++ b/pkg/sql/colexec/range_stats.go
@@ -20,6 +20,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/colmem"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/util/json"
+	"github.com/cockroachdb/errors"
 )
 
 type rangeStatsOperator struct {
@@ -90,7 +91,13 @@ func (r *rangeStatsOperator) Next() coldata.Batch {
 	r.allocator.PerformOperation(
 		[]coldata.Vec{output},
 		func() {
+			// TODO(yuzefovich): consider reusing these slices across
+			// iterations.
 			keys := make([]roachpb.Key, 0, batch.Length())
+			// keysOutputIdx is a 1-to-1 mapping with keys which stores the
+			// position in the output vector that the results of the range stats
+			// lookup for the corresponding key should be put into.
+			keysOutputIdx := make([]int, 0, batch.Length())
 			if inSel == nil {
 				for i := 0; i < batch.Length(); i++ {
 					if inNulls.MaybeHasNulls() && inNulls.NullAt(i) {
@@ -98,6 +105,7 @@ func (r *rangeStatsOperator) Next() coldata.Batch {
 						continue
 					}
 					keys = appendKey(keys, inBytes, i)
+					keysOutputIdx = append(keysOutputIdx, i)
 				}
 			} else {
 				for _, idx := range inSel {
@@ -106,10 +114,15 @@ func (r *rangeStatsOperator) Next() coldata.Batch {
 						continue
 					}
 					keys = appendKey(keys, inBytes, idx)
+					keysOutputIdx = append(keysOutputIdx, idx)
 				}
 			}
 			if inNulls.MaybeHasNulls() {
 				output.Nulls().Copy(inNulls)
+			}
+			if len(keys) == 0 {
+				// All keys were NULL.
+				return
 			}
 			// TODO(ajwerner): Reserve memory for the responses. We know they'll
 			// at least, on average, contain keys so it'll be 2x the size of the
@@ -118,8 +131,13 @@ func (r *rangeStatsOperator) Next() coldata.Batch {
 			if err != nil {
 				colexecerror.ExpectedError(err)
 			}
-			readResponse := func(resultIndex, outputIndex int) {
-				jsonStr, err := gojson.Marshal(&res[resultIndex].MVCCStats)
+			if len(res) != len(keys) {
+				colexecerror.InternalError(errors.AssertionFailedf(
+					"unexpected number of RangeStats responses %d: %d expected", len(res), len(keys),
+				))
+			}
+			for i, outputIdx := range keysOutputIdx {
+				jsonStr, err := gojson.Marshal(&res[i].MVCCStats)
 				if err != nil {
 					colexecerror.ExpectedError(err)
 				}
@@ -127,16 +145,7 @@ func (r *rangeStatsOperator) Next() coldata.Batch {
 				if err != nil {
 					colexecerror.ExpectedError(err)
 				}
-				jsonOutput.Set(outputIndex, jsonDatum)
-			}
-			if inSel != nil {
-				for i, s := range inSel {
-					readResponse(i, s)
-				}
-			} else {
-				for i := range res {
-					readResponse(i, i)
-				}
+				jsonOutput.Set(outputIdx, jsonDatum)
 			}
 		})
 	return batch

